Checkpoint (#1)
ekr/ppm-specification: draft-pda-protocol.md

protocol* is to compute `y = F(x_1, ..., x_n)` for some aggregation
function `F` while revealing nothing else about the measurements.

This protocol is extensible and allows for the addition of new
cryptographic schemes that compute new functions. The current
version supports two schemes:

* Prio {{CB17}}, which allows for aggregate statistics such as
  sum, mean, histograms, etc. over a single value.

* Heavy Hitters {{BBCp21}}, which allows for finding the most
  common strings among a collection of clients (e.g., the
  URL of their home page) as well as counting the number of
  clients that hold a given string.

This protocol is designed to work with schemes that use secret
sharing. Rather than send its input in the clear, each client splits
its measurements into a sequence of *shares* and sends a share to each
of the aggregators. This provides two important properties:

* It's impossible to deduce the measurement without knowing *all* of the shares.

* It allows the aggregators to compute the final output by first
   aggregating up their measurements shares locally, then combining
   the results to obtain the final output.

> 

Reviewer: ```suggestion
Aggregators also have the option of validating shares, as discussed in {{validating-inputs}}.
```

Author: The option? Isn't it required effectively?

Reviewer: There were some deployments that wanted to use this without actually verifying any proofs 🤷 , relying on client attestation and authentication of some form to ensure the data is trusted.

## System Architecture

The overall system architecture is shown in {{pa-topology}}.

~~~~
                    +------------+
                    |            |
+--------+          |   Helper   |
|        |          |            |
| Client +----+     +-----^------+
|        |    |           |
+--------+    |           |
              |           |
+--------+    |     +-----v------+         +-----------+
|        |    +----->            |         |           |

> | Client +---------->   Leader   +---------> Collector |

Reviewer: ```suggestion
| Client +---------->   Leader   <---------> Collector |
```

|        |    +----->            |         |           |
+--------+    |     +-----^------+         +-----------+
              |           |
+--------+    |           |
|        |    |           |
| Client +----+     +-----V------+
|        |          |            |
+--------+          |   Helper   |
                    |            |
                    +------------+
~~~~
{: #pa-topology title="System Architecture"}

The main participants in the protocol are as follows:

Collector:
: The entity which wants to take the measurement and ultimately receives
  the results. Any given measurement will have a single collector.

Client(s):

> : The endpoints which directly take the measurement(s) and report it to the

Reviewer: ```suggestion
: The endpoints which directly take the measurement(s) and report them to the
```

  PPM system. In order to provide reasonable levels of privacy, there

>   must be a large number of clients.

Reviewer: ```suggestion
  must be a large number of clients.
```

Aggregator:
: An endpoint which receives report shares. Each aggregator works with the
  other aggregators to compute the final aggregate. This protocol defines

>   two types of aggregators: a Leader and a set of Helpers.

Reviewer: ```suggestion
  two types of aggregators: Leaders and Helpers. For each measurement,
  there is a single leader and helper.
```

Leader:
: The leader is responsible for coordinating the protocol. It receives

>   the encrypted shares, distributes them to the helpers and orchestrates

Reviewer: ```suggestion
  the encrypted shares, distributes them to the helpers, and orchestrates
```

>   the process of computing the final measurement.

Reviewer: ```suggestion
  the process of computing the final measurement with the Collector.
```

Author: I see what you want, but I think this is a bit confusing as written
"maybe as requested by the Collector"

Reviewer: Yeah, that's better

Helper:
: Helpers are responsible for executing the protocol as instructed
  by the leader. The protocol is designed so that helpers can be relatively
  lightweight, with most of the state held at the leader.
{:br}

In order to take a measurement, the collector first decides on the
measurement parameters, including:

> * The values to be measured
> * The statistic to be computed (e.g., sum, mean, etc.)
> * The PPM scheme to use. This is to some extent dictated by the previous
>   choices.

Reviewer: ```suggestion
* The values to be measured;
* The statistic to be computed (e.g., sum, mean, etc.); 
* The set of aggregators and necessary cryptographic keying material to use; and
* The PPM scheme to use. This is to some extent dictated by the previous
  choices.
```

These parameters are distributed out of band to the clients and to
the aggregators.

During the duration of the measurement, each client records its own

> value(s), packages them up into a report and sends them to the leader.

Reviewer: ```suggestion
value(s), packages them up into a report, and sends them to the leader.
```

Each share is separately encrypted for each aggregator so that even
though they pass through the leader, it is unable to see or modify
them. Depending on the measurement, the client may only send one
report or may send many reports over time.

The leader distributes the shares to the helpers and orchestrates
the process of verifying them (see {{validating-inputs}})
and assembling them into a final measurement for the collector.
Depending on the PPM scheme, it may be possible to incrementally
process each report as it comes in, or may be necesary to wait

> until all the required shares are received. This protocol is
> compatible with both a "push" mode in which the leader computes
> the measurements at predetermined points and sends the results
> to the collector and a "pull" mode in which the collector tells
> the leader when to perform the aggregation and return the results.

Reviewer: ```suggestion
until all the required shares are received. 

This protocol is compatible with both a "push" deployment mode in which the 
leader computes the measurements at predetermined points and sends the results
to the collector and a "pull" mode in which the collector tells
the leader when to perform the aggregation and return the results.
```

Reviewer: Tried to clarify that modes are a deployment choice, and not something baked into the protocol.

"extension_type" indicates the type of extension, and "extension_data" contains
information specific to the extension.

## Verifying and Aggregating Reports {#pa-aggregate}

Once a set of clients have uploaded their reports to the leader, the leader
can send them to the aggregators to be verified and aggregated. In order
to enable the system to handle very large batches of reports, this process
can be performed incrementally. To aggregate a set of reports,
the leader sends an AggregateReq to the helper containing those report

> shares. The helper then processes the (verifying

Reviewer: ```suggestion
shares. The helper then processes them (verifying
```
